File: Solution/src/com/solution/tree/ValidBST.java
package com.solution.tree;

import com.solution.main.BTreePrinter;
import com.solution.main.ReadInput;
import com.solution.model.Tree;

public class ValidBST {

	public void run() {
		Tree t = ReadInput.getTree("Tree");

		BTreePrinter.printNode(t);

		boolean isValid = doRun(t);

		if (isValid) {
			System.out.println("yes");
		} else {
			System.out.println("no");
		}

	}

	private boolean doRun(Tree t) {

		if (t == null) {
			return true;
		}

		// leaf node
		if (t.getLeft() == null && t.getRight() == null) {
			return true;
		}

		int data = (int) t.getData();
		Tree left = t.getLeft();
		Tree right = t.getRight();

		if (null != left && null != right) {
			if (data > (int) left.getData() && data < (int) right.getData()) {
				return doRun(t.getLeft()) && doRun(t.getRight());
			}
		} else if (left != null) {
			if (data > (int) left.getData()) {
				return doRun(left);
			}
		} else if (right != null) {
			if (data < (int) right.getData()) {
				return doRun(right);
			}
		}
		
		return false;
	}
}

File: src/main/kotlin/com/cssolutions/dsalgo/tree/ValidBST.kt
package com.cssolutions.dsalgo.tree

import com.cssolutions.dsalgo.main.BTreePrinter.printNode
import com.cssolutions.dsalgo.main.ReadInput.getTree
import com.cssolutions.dsalgo.model.Tree

class ValidBST {
    fun run() {
        val t = getTree("Tree")
        printNode<Comparable<*>>(t)
        val isValid = doRun(t)
        if (isValid) {
            println("yes")
        } else {
            println("no")
        }
    }

    private fun doRun(t: Tree?): Boolean {
        if (t == null) {
            return true
        }
        // leaf node
        if (t.left == null && t.right == null) {
            return true
        }
        val data = t.data as Int
        val left = t.left
        val right = t.right
        if (null != left && null != right) {
            if (data > left.data as Int && data < right.data as Int) {
                return doRun(t.left) && doRun(t.right)
            }
        } else if (left != null) {
            if (data > left.data as Int) {
                return doRun(left)
            }
        } else if (right != null) {
            if (data < right.data as Int) {
                return doRun(right)
            }
        }
        return false
    }
}